src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2437,39 +2585,86 @@ public override BoundNode VisitTupleBinaryOperator(BoundTupleBinaryOperator node
                     new FormattedSymbol(method, SymbolDisplayFormat.MinimallyQualifiedFormat),
                     delegateType);
             }
 
             int count = Math.Min(invoke.ParameterCount, method.ParameterCount);
-
             for (int i = 0; i < count; i++)
             {
-                if (IsNullabilityMismatch(invoke.Parameters[i].Type, method.Parameters[i].Type))
+                var invokeParameter = invoke.Parameters[i];
+                var methodParameter = method.Parameters[i];
+                if (IsNullabilityMismatch(invokeParameter.Type, methodParameter.Type, requireIdentity: invokeParameter.RefKind != RefKind.None))
                 {
                     ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, syntax,
-                        new FormattedSymbol(method.Parameters[i], SymbolDisplayFormat.ShortFormat),
+                        new FormattedSymbol(methodParameter, SymbolDisplayFormat.ShortFormat),
                         new FormattedSymbol(method, SymbolDisplayFormat.MinimallyQualifiedFormat),
                         delegateType);
                 }
             }
         }
 
+        private void ReportNullabilityMismatchWithTargetDelegate(SyntaxNode syntax, NamedTypeSymbol delegateType, UnboundLambda unboundLambda)
+        {
+            if (!unboundLambda.HasExplicitlyTypedParameterList)
+            {
+                return;
+            }
+
+            MethodSymbol invoke = delegateType?.DelegateInvokeMethod;
+            if (invoke is null)
+            {
+                return;
+            }
+
+            int count = Math.Min(invoke.ParameterCount, unboundLambda.ParameterCount);
+            for (int i = 0; i < count; i++)
+            {
+                var invokeParameter = invoke.Parameters[i];
+                // Parameter nullability is expected to match exactly. This corresponds to the behavior of initial binding.
+                //    Action<string> x = (object o) => { }; // error CS1661: Cannot convert lambda expression to delegate type 'Action<string>' because the parameter types do not match the delegate parameter types
+                //    Action<object> y = (object? o) => { }; // warning CS8622: Nullability of reference types in type of parameter 'o' of 'lambda expression' doesn't match the target delegate 'Action<object>'.
+                // PROTOTYPE(NullableReferenceTypes): Consider relaxing and allow implicit conversions of nullability.
+                // (Compare with method group conversions which pass `requireIdentity: false`.)
+                if (IsNullabilityMismatch(invokeParameter.Type, unboundLambda.ParameterType(i), requireIdentity: true))
+                {
+                    ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, syntax,
```

Comment:
>syntax [](start = 122, length = 6)

Can the syntax node we report this on be more precise (i.e. the lambda parameter)?

Reply:
Added comment.

---
In reply to: [193913531](https://github.com/dotnet/roslyn/pull/27283#discussion_r193913531) [](ancestors = 193913531)